pkg/sync/sync.go

```diff
@@ -25,41 +25,69 @@ func NewMutexMap[K comparable]() *MutexMap[K] {
 }
 
 type mutexMapEntry struct {
-	mu    sync.Mutex
-	count int
+	mu         sync.Mutex
+	readerCnt  int
+	writerCnt  int
+	readerDone *sync.Cond
 }
 
-// Lock locks the given key.
+// Lock locks the given key for writing.
 func (m *MutexMap[K]) Lock(key K) {
 	m.mu.Lock()
-	entry := m.entries[key]
-	if entry == nil {
-		entry = &mutexMapEntry{}
-		m.entries[key] = entry
-	}
-	entry.count++
+	entry := m.getOrCreateEntry(key)
+	entry.writerCnt++
 	m.mu.Unlock()
-
 	entry.mu.Lock()
+	for entry.readerCnt > 0 {
+		entry.readerDone.Wait()
+	}
 }
 
-// Unlock unlocks the given key.
+// Unlock unlocks the given key for writing.
 func (m *MutexMap[K]) Unlock(key K) {
```

Comment:
why we cannot use existing solution?

Reply:
If we try to use existing solution here, it would lead to deadlock due to the same reason you mentioned. Hence, I have used conditional synchronization..

Reply:
Using the existing solution now.